app/assets/javascripts/discourse/app/lib/uppy-chunked-upload.js

```diff
@@ -0,0 +1,339 @@
+import { Promise } from "rsvp";
+import delay from "@uppy/utils/lib/delay";
+import {
+  AbortController,
+  createAbortError,
+} from "@uppy/utils/lib/AbortController";
+
+const MB = 1024 * 1024;
+
+const defaultOptions = {
+  limit: 5,
+  retryDelays: [0, 1000, 3000, 5000],
+  getChunkSize() {
+    return 5 * MB;
+  },
+  onStart() {},
+  onProgress() {},
+  onChunkComplete() {},
+  onSuccess() {},
+  onError(err) {
+    throw err;
+  },
+};
+
+/**
+ * Used mainly as a replacement for Resumable.js, using code cribbed from
+ * uppy's S3 Multipart class, which we mainly use the chunking algorithm
+ * and retry/abort functions of. The _buildFormData function is the one
+ * which shapes the data into the same parameters as Resumable.js used.
+ *
+ * See the UppyChunkedUploader class for the uppy uploader plugin which
+ * uses UppyChunkedUpload.
+ */
+export default class UppyChunkedUpload {
+  constructor(file, options) {
+    this.options = {
+      ...defaultOptions,
+      ...options,
+    };
+    this.file = file;
+
+    if (!this.options.getChunkSize) {
+      this.options.getChunkSize = defaultOptions.getChunkSize;
+      this.chunkSize = this.options.getChunkSize(this.file);
+    }
+
+    this.abortController = new AbortController();
+    this.chunksInProgress = 0;
+    this.chunks = null;
+    this.chunkState = null;
+
+    this._initChunks();
```

Comment:
Supernit: if you're using _initChunks, I guess the top few chunk related setup can go there too?

Reply:
Good catch!